aten/src/ATen/native/cuda/GridSampler.cu

```diff
@@ -0,0 +1,782 @@
+#include "ATen/ATen.h"
+#include "ATen/native/GridSampler.h"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/detail/TensorInfo.cuh"
+#include "ATen/cuda/detail/IndexUtils.cuh"
+#include "ATen/cuda/detail/KernelUtils.h"
+
+namespace at { namespace native {
+
+using namespace at::cuda::detail;
+
+using at::native::detail::GridSamplerInterpolation;
+using at::native::detail::GridSamplerPadding;
+
+namespace {
+	static __forceinline__ __device__
+	int clip_coordinates(int in, int clip_limit) {
+	  return ::min(clip_limit - 1, ::max(in, static_cast<int>(0)));
+	}
+
+	static __forceinline__ __device__
+	bool within_bounds_2d(int h, int w, int H, int W) {
+	  return h >= 0 && h < H && w >= 0 && w < W;
+	}
+
+	static __forceinline__ __device__
+	bool within_bounds_3d(int d, int h, int w, int D, int H, int W) {
+	  return d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W;
+	}
+
+  template<typename scalar_t>
+	static __forceinline__ __device__
+  void safe_add_2d(scalar_t *data, int h, int w,
+                   int sH, int sW, int H, int W,
+                   scalar_t delta) {
+    if (h >= 0 && h < H && w >= 0 && w < W) {
```

Comment:
Any particular reason you didn't use `within_bounds` here?

Reply:
ah good point. I'll use `within_bounds`.